src/xmipp/libraries/reconstruction/subtract_projection.cpp

```diff
@@ -84,287 +86,283 @@
-		extractPhaseProj(PFourierPhase);
-		for (int n=0; n<iter; ++n) {
-			runIteration();
-			if (cutFreq!=0) {
-				Filter2.generateMask(Pctf());
-				Filter2.do_generate_3dmask=true;
-				Filter2.applyMaskSpace(Pctf());
+	// Initialize Fourier projectors
+	double cutFreq = sampling/maxResol;
+	projector = std::make_unique<FourierProjector>(V(), padFourier, cutFreq, xmipp_transformation::BSPLINE3);
+	std::unique_ptr<FourierProjector> projectorMask;
+	projectorMask = std::make_unique<FourierProjector>(vM(), padFourier, cutFreq, xmipp_transformation::BSPLINE3);
+	// Read first particle
+	const auto sizeI = (int)XSIZE(I());
+	processParticle(1, sizeI, transformerP, transformerI);
+	const MultidimArray<double> &mPctf = Pctf();
+	// Construct frequencies image
+	MultidimArray<int> wi;
+	wi.initZeros(PFourier);
+	Matrix1D<double> w(2); 	
+	for (int i=0; i<YSIZE(wi); i++) {
+		FFT_IDX2DIGFREQ(i,YSIZE(mPctf),YY(w)) 
+		for (int j=0; j<XSIZE(wi); j++)  {
+			FFT_IDX2DIGFREQ(j,XSIZE(mPctf),XX(w))
+			DIRECT_A2D_ELEM(wi,i,j) = (int)round((sqrt(YY(w)*YY(w) + XX(w)*XX(w))) * (int)XSIZE(mPctf)); 
+		}
+	}
+	int maxwiIdx;
+	if (limitfreq == 0)
+		maxwiIdx = (int)XSIZE(wi); 
+	else
+		DIGFREQ2FFT_IDX(cutFreq, (int)YSIZE(PFourier), maxwiIdx)
+
+	// Declare complex structures that will be used in the loop
+	FourierTransformer transformerIiM;
+	FourierTransformer transformerPiM;
+	MultidimArray< std::complex<double> > IiMFourier;
+	MultidimArray< std::complex<double> > PiMFourier;
+	
+	// For each particle in metadata:
+    for (size_t i = 1; i <= mdParticles.size(); ++i) {  
+     	// Project volume and process projections 
+		processParticle(i, sizeI, transformerP, transformerI);
+		// Build projected and final masks
+		if (fnMask.isEmpty() || fmaskWidth == -1) {
+			Mfinal().initZeros(P());
+			iM = invertMask(Mfinal);
+			Mfinal = iM;		
 			}
+		else {
+			projectVolume(*projectorMask, Pmask, sizeI, sizeI, part_angles.rot, part_angles.tilt, part_angles.psi, ctfImage);	
+			M = binarizeMask(Pmask);
+			Mfinal().initZeros(M());
+			auto fmaskWidth_px = fmaskWidth/(int)sampling;
+			dilate2D(M(), Mfinal(), 8, 0, fmaskWidth_px); 
+			FilterG.applyMaskSpace(M());
+			iM = invertMask(M);
 		}
-		Image<double> IFiltered;
-    	I.read(fnImage);
-		IFiltered() = I();
-		if (cutFreq!=0)
-			Filter2.applyMaskSpace(IFiltered());
-    	projectVolume(mask(), Pmask, (int)XSIZE(I()), (int)XSIZE(I()), angles.rot, angles.tilt, angles.psi, &roffset);
-    	Pmaskctf = applyCTF(row, Pmask);
-    	Pmaskctf = thresholdMask(Pmaskctf);
-		PmaskInv = invertMask(Pmaskctf);
-    	FilterG.w1=sigma;
-		FilterG.applyMaskSpace(Pmaskctf());
+		// Compute estimation images: IiM = I*iM and PiM = P*iM	
+		IiMFourier = computeEstimationImage(I(), iM(), transformerIiM);
+		PiMFourier = computeEstimationImage(P(), iM(), transformerPiM);
+
+		// Estimate transformation with model of order 0: T(w) = beta00 and model of order 1: T(w) = beta01 + beta1*w
+		MultidimArray<double> num0;
```

Comment:
Just a suggestion, could it run with floats instead of double?
You define several double matrixes, and we could improve the times with floats (not sure)

Reply:
I've thought about it because some variables do not need such a big precision, however I prefer to keep double for images and fourier transforms and thus, when operating with images and FTs values and float variables I would need to do a cast to doubles again for not loosing precission (and also for passing the sonar cloud analysis) thus I think it is better to mantain all to doubles